### PR TITLE
Add etcd-empty-dir-cleanup.yml cleanup inspired from kubernetes project

### PR DESCRIPTION
**Pull request by gbergere**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,23 @@
 FROM alpine:3.6
 
+ENV ETCD_VERSION 3.1.4
+ENV KUBE_VERSION 1.6.2
```
Comment (adrianchifor) [inline Dockerfile]: We can move this to 1.7.8

File: Dockerfile [hunk 2]
```diff
@@ -1,13 +1,23 @@
...
-COPY docker-clean.sh k8s-clean.sh /bin/
+RUN cd /tmp \
+    && curl -OL https://github.com/coreos/etcd/releases/download/v${ETCD_VERSION}/etcd-v${ETCD_VERSION}-linux-amd64.tar.gz \ 
+    && tar zxf etcd-v${ETCD_VERSION}-linux-amd64.tar.gz \
+    && cp etcd-v${ETCD_VERSION}-linux-amd64/etcdctl /usr/local/bin/etcdctl \
+    && rm -rf etcd-v${ETCD_VERSION}-linux-amd64* \
+    && chmod +x /usr/local/bin/etcdctl
+
+COPY docker-clean.sh k8s-clean.sh etcd-empty-dir-cleanup.sh /bin/
 RUN chmod +x /bin/docker-clean.sh && chmod +x /bin/k8s-clean.sh
```
Comment (adrianchifor) [inline Dockerfile]: Add `etcd-empty-dir-cleanup.sh` to this line as well

File: etcd-empty-dir-cleanup.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    do
+      cleanup_empty_dirs ${SUBDIR}
+    done
+  else
+    echo "Removing empty key $1 ..."
+    etcdctl rmdir $1
+  fi
+}
+
+while true
```
Comment (adrianchifor) [inline etcd-empty-dir-cleanup.sh]: We can remove the while loop as this runs only once

File: etcd-empty-dir-cleanup.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  namespace: kube-system
+spec:
+  schedule: "0 2 * * *"
+  concurrencyPolicy: "Forbid"
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          name: etcd-empty-dir-cleanup
+          annotations:
```
Comment (adrianchifor) [inline etcd-empty-dir-cleanup.yml]: We can remove  all of `annotations`, doesn't work on k8s 1.7

File: Dockerfile (hunk 2 above)
Comment (gbergere) [reply Dockerfile]: What do you mean ? It's already line 20.

File: etcd-empty-dir-cleanup.sh (hunk 3 above)
Comment (gbergere) [reply etcd-empty-dir-cleanup.sh]: Fair point, silly me.

File: etcd-empty-dir-cleanup.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+              [
+                {
+                  "key": "dedicated",
+                  "operator": "Equal",
+                  "value": "master",
+                  "effect": "NoSchedule"
+                }
+              ]
+        spec:
+          nodeSelector:
```
Comment (adrianchifor) [inline etcd-empty-dir-cleanup.yml]: You can use the master toleration as such:
```
nodeSelector:
  kubernetes.io/role: master
tolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
```

File: etcd-empty-dir-cleanup.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            kubernetes.io/role: master
+          restartPolicy: OnFailure
+          serviceAccountName: k8s-cleanup
+          containers:
+          - name: k8s-cleanup
+            image: onfido/k8s-cleanup
+            command: ["/bin/bash", "/bin/k8s-clean.sh"]
```
Comment (adrianchifor) [inline etcd-empty-dir-cleanup.yml]: `/bin/etcd-empty-dir-cleanup.sh` right?

File: README.md [hunk 7]
```diff
@@ -1,6 +1,9 @@
 ## k8s-cleanup
 
-Cleans up exited containers and dangling images/volumes running as a DaemonSet (`deploy-ds.yml`) and cleans up old replica sets and finished jobs as a CronJob (`deploy-cron.yml`).
+Here are 3 cleanups you can apply on your kubernetes cluster:
+* Cleans up exited containers and dangling images/volumes running as a DaemonSet (`deploy-ds.yml`).
```
Comment (adrianchifor) [inline README.md]: If you're calling the last `etcd-empty-dir-cleanup.yml`, might be worth renaming `deploy-ds.yml` and `deploy-cron.yml` to `docker-clean.yml` and `k8s-clean.yml` ?

File: Dockerfile (hunk 2 above)
Comment (adrianchifor) [reply Dockerfile]: It's copied but not allowed to execute. We might not need those actually as we run them as `/bin/bash <script>`

Comment (gbergere) [reply Dockerfile]: Oh yes ! 👍

File: etcd-empty-dir-cleanup.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  if [ "$(etcdctl ls $1)" ]; then
+    for SUBDIR in $(etcdctl ls -p $1 | grep "/$")
+    do
+      cleanup_empty_dirs ${SUBDIR}
+    done
+  else
+    echo "Removing empty key $1 ..."
+    etcdctl rmdir $1
+  fi
+}
```
Comment (adrianchifor) [inline etcd-empty-dir-cleanup.sh:30]: I think you removed the function call as well 😄

Comment (gbergere) [reply etcd-empty-dir-cleanup.sh:30]: Solved :)

**Pull request merged**